### ISSUE TITLE
cuda topk dim=0 crash on small tensor in two lines of code

### ISSUE THREAD

**Issue by davidbau** [pytorch/pytorch#4490]
Repro:
```
import torch
a = torch.randn(25, 1, 8).cuda()
b = torch.randn(25, 1, 9).cuda()
print('This one works')
print(a.topk(3, dim=0)[1])
print('This one crashes')
print(b.topk(3, dim=0)[1])
```

Expect two small tensors of small integers to be printed.  Instead, only the first case works. In the second case, you get a series of failed assertions starting with this:

```
pytorch/aten/src/THC/THCTensorTopK.c
uh:251: DataType findPattern(DataType *, DataType *, IndexType, IndexType, BitDa
taType, BitDataType) [with DataType = float, BitDataType = unsigned int, IndexTy
pe = unsigned int]: block: [46,0,0], thread: [0,0,0] Assertion `false` failed.
```

I wonder if it is related to #3959

**Comment by albanD** [comment]
@davidbau is this still an outstanding issue?
I cannot reproduce this with an install from the master branch.

**Comment by davidbau** [comment]
The size which causes failure doesn't seem deterministic.  Using a pretty recent build (a few days old), I now have to increase the "9" to a larger number to make it fail.  Sometimes it will fail at 10, sometimes at 11.

Here's an updated case that seems to fail reliably at some size, for me:

```
import torch
for size in range(8, 16):
    a = torch.randn(25, 1, size).cuda()
    print('Testing size: %d' % size)
    print(a.topk(3, dim=0)[1])
    print('OK...')
```
